File: apps/web/src/lib/departmentExecutionOverview.ts
import type {
  CgCurrentFocusType,
  DepartmentExecutionAnchorState,
  DepartmentExecutionLastUpdatedSource,
  DepartmentExecutionVersionScope,
} from "@intent-core/contracts";

import type { StatusBadgeStatus } from "@/design";

/** Step 9B-3: honest, human-readable copy for each real
 * `DepartmentExecutionAnchorState` value -- never the raw enum, and
 * never implying "confirmed" for anything short of a real confirmed
 * revision. */
const STATE_LABEL: Record<DepartmentExecutionAnchorState, string> = {
  none: "No Execution Anchor yet",
  draft: "Draft awaiting CG completion",
  awaiting_confirmation: "Awaiting CG confirmation",
  confirmed: "Confirmed",
  rejected: "Rejected",
};

const STATE_BADGE_STATUS: Record<
  DepartmentExecutionAnchorState,
  StatusBadgeStatus
> = {
  none: "neutral",
  draft: "attention",
  awaiting_confirmation: "attention",
  confirmed: "confirmed",
  rejected: "blocking",
};

export function executionAnchorStateLabel(
  state: DepartmentExecutionAnchorState,
  revisionNumber: number | null,
): string {
  const label = STATE_LABEL[state];
  if (state === "confirmed" && revisionNumber !== null) {
    return `${label} (Revision ${revisionNumber})`;
  }
  return label;
}

export function executionAnchorStateBadgeStatus(
  state: DepartmentExecutionAnchorState,
): StatusBadgeStatus {
  return STATE_BADGE_STATUS[state];
}

/** Step 9B-3: which real source object contributed a row's
 * `last_updated_at` -- rendered as a short, human-readable phrase, never
 * the raw internal discriminator string. */
const LAST_UPDATED_SOURCE_LABEL: Record<
  DepartmentExecutionLastUpdatedSource,
  string
> = {
  task_created: "Task created",
  execution_anchor_revision: "Execution Anchor updated",
  version: "Production Version recorded",
  dependency: "Dependency recorded",
  escalation: "Escalation recorded",
  alignment_assessment: "Alignment Assessment generated",
};

export function lastUpdatedSourceLabel(
  source: DepartmentExecutionLastUpdatedSource,
): string {
  return LAST_UPDATED_SOURCE_LABEL[source];
}

/** Step 9B-3 owner-validation correction: CG's own Current-focus copy
 * (`cg_inbox.current_focus`) is written in CG's second person ("your
 * confirmation", "your interpretation") for CG's own Review Inbox. That
 * copy must never be rendered unqualified on a VFX page -- it would read
 * as if it were addressed to the VFX Supervisor, and would visually
 * contradict a separately-shown advisory alignment concern or the "no
 * open escalation" line on the same row. This is role-explicit,
 * role-neutral copy keyed only on the real `current_focus_type`
 * discriminator -- never a second focus-ranking algorithm, only a
 * different rendering of the same real state. */
const CG_TASK_FOCUS_LABEL: Record<CgCurrentFocusType, string> = {
  execution_anchor_gate_pending:
    "CG task focus: Execution Anchor draft awaiting CG confirmation.",
  execution_anchor_draft_needs_review:
    "CG task focus: Execution Anchor draft in progress.",
  dependency_needs_attention:
    "CG task focus: an unresolved dependency needs CG interpretation.",
  version_review_available:
    "CG task focus: a Production Version is ready for CG review.",
  none: "No current CG action is required for this Task.",
};

export function cgTaskFocusLabel(focusType: CgCurrentFocusType): string {
  return CG_TASK_FOCUS_LABEL[focusType];
}

/** Step 9B-3 owner-validation correction: a Task-scoped Version
 * (`Version.task_id === task.task_id`) must be visibly distinguished
 * from the Step 8C-6/8C-7 nullable-`task_id` Shot-level fallback -- the
 * latter is not explicitly linked to this Task in ICAS, even when it is
 * the only Version returned. */
const VERSION_SCOPE_LABEL: Record<DepartmentExecutionVersionScope, string> = {
  task: "Task-linked Version",
  shot_unscoped:
    "Shot-level Version fallback — not linked to this Task in ICAS",
};

export function latestVersionScopeLabel(
  scope: DepartmentExecutionVersionScope | null,
): string | null {
  return scope === null ? null : VERSION_SCOPE_LABEL[scope];
}

File: packages/contracts/python/src/intent_core_contracts/api/department_execution_overview.py
"""Request/response schemas for the VFX Department Execution Overview
read model (Step 9B-3; docs/step-9/05_STEP_9B3_DEPARTMENT_EXECUTION_OVERVIEW.md).

Read-only, Shot-scoped, VFX-Supervisor-only. Composes real, already-
persisted Task/ExecutionAnchor/ExecutionAnchorRevision/HumanGate/Version/
TaskDependency/CrossRoleAssessment/IntentSignal state into one row per
real Task under a Shot -- the same "one row per real object, several
already-real fields projected together" pattern already established by
`VfxInboxItemRead`/`CgInboxItemRead`. No new authoritative table, no new
mutation. Every row is matched by its real `Task.id` -- never by a
department-name string (docs/step-9/02_STEP_9A_...md §8).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from intent_core_contracts.api.dependencies import TaskDependencySeverity
from intent_core_contracts.api.production_context import RecordSource
from intent_core_contracts.api.vfx_inbox import AttentionLevel

DepartmentExecutionAnchorState = Literal[
    "none",
    "draft",
    "awaiting_confirmation",
    "confirmed",
    "rejected",
]

DepartmentExecutionLastUpdatedSource = Literal[
    "task_created",
    "execution_anchor_revision",
    "version",
    "dependency",
    "escalation",
    "alignment_assessment",
]


class DepartmentExecutionTaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=False)

    task_id: UUID
    task_name: str
    department: str | None
    task_source: RecordSource

    # The Task's current Execution Anchor state, using the exact
    # lifecycle rule the service documents: `confirmed` only from a real
    # `active_revision_id` whose own status is `confirmed`; `awaiting_
    # confirmation` only when a real pending HumanGate exists; `draft`
    # only for an un-gated draft revision; `rejected` only when the
    # latest real revision's own status is `rejected`; a superseded
    # revision is never selected as current under any state.
    execution_anchor_state: DepartmentExecutionAnchorState
    execution_anchor_revision_number: int | None
    execution_anchor_summary: str | None

    # Latest Production Version relevant to this Task: Task-scoped first
    # (`Version.task_id == task_id`), falling back to a Shot-level manual
    # Version (`Version.task_id IS NULL`) per the existing Step 8C-6/8C-7
    # nullable-task_id compatibility rule -- never a Version explicitly
    # linked to a *different* Task. Ordered by `source_created_at ??
    # created_at`.
    latest_version_id: UUID | None
    latest_version_name: str | None
    latest_version_number: int | None
    latest_version_source: RecordSource | None

    # The Task's existing, real, already-tested Current focus (the same
    # derivation CG's own Review Inbox uses, `cg_inbox.current_focus`) --
    # never re-derived by a second ranking algorithm.
    current_focus_title: str
    current_focus_actionable: bool

    # Real, open TaskDependency rows (`kind in ("dependency", "conflict")`)
    # only -- an escalation is reported separately below, and a resolved
    # Dependency is never counted or shown here as open.
    open_dependency_count: int
    top_open_dependency_description: str | None
    top_open_dependency_severity: TaskDependencySeverity | None

    # The Task's own latest CrossRoleAssessment's IntentSignal, if any --
    # always advisory (Agent-derived), never a Human Decision. `None`
    # means no Assessment has been generated for this Task yet -- never
    # rendered as "confirmed aligned."
    alignment_concern_summary: str | None
    alignment_concern_attention_level: AttentionLevel | None

    # A real, open TaskDependency(kind="escalation",
    # escalated_to_role="vfx_supervisor") targeting this Task -- never
    # inferred from an Agent recommendation or a high Intent Signal alone.
    open_escalation: bool
    open_escalation_summary: str | None

    # Deterministic maximum of every real, included source object's own
    # timestamp above -- never the request time. `last_updated_source`
    # names exactly which one won (ties broken by a fixed field order,
    # never arbitrarily).
    last_updated_at: datetime
    last_updated_source: DepartmentExecutionLastUpdatedSource


class DepartmentExecutionOverviewRead(BaseModel):
    model_config = ConfigDict(from_attributes=False)

    shot_id: UUID
    tasks: list[DepartmentExecutionTaskRead]
    generated_at: datetime
